File: ios/Runner/EspProvisioningHandler.swift
import Foundation
import Flutter
import NetworkExtension

/// Platform channel handler for ESP Wi-Fi provisioning functionality on iOS.
///
/// This provides native iOS implementations for:
/// - ESP SmartConfig (ESPTouch/ESPTouchV2)
/// - BLE-based ESP provisioning
/// - SoftAP-based ESP provisioning
/// - Wi-Fi connection management
///
/// Note: Full ESP provisioning requires native libraries from Espressif.
/// This implementation provides the channel interface and basic Wi-Fi connectivity.
/// For full ESP provisioning support, integrate the Espressif provisioning library:
/// https://github.com/espressif/esp-idf-provisioning-ios
class EspProvisioningHandler: NSObject {
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        // ESP SmartConfig methods
        case "startSmartConfig":
            startSmartConfig(call, result: result)
        case "stopSmartConfig":
            stopSmartConfig(result: result)
            
        // BLE Provisioning methods
        case "scanBleDevices":
            scanBleDevices(call, result: result)
        case "scanWifiNetworksBle":
            scanWifiNetworksBle(call, result: result)
        case "provisionWifiBle":
            provisionWifiBle(call, result: result)
            
        // SoftAP Provisioning methods
        case "startSoftApSession":
            startSoftApSession(call, result: result)
        case "scanWifiNetworksSoftAp":
            scanWifiNetworksSoftAp(result: result)
        case "sendWifiConfigSoftAp":
            sendWifiConfigSoftAp(call, result: result)
        case "applyWifiConfigSoftAp":
            applyWifiConfigSoftAp(result: result)
        case "getStatusSoftAp":
            getStatusSoftAp(result: result)
        case "sendReceiveCustomDataSoftAp":
            sendReceiveCustomDataSoftAp(call, result: result)
        case "disposeSoftApSession":
            disposeSoftApSession(result: result)
            
        // Wi-Fi Connect methods
        case "connectToWifi":
            connectToWifi(call, result: result)
        case "connectToSecureWifi":
            connectToSecureWifi(call, result: result)
        case "disconnectFromWifi":
            disconnectFromWifi(result: result)
            
        // Utility
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - SmartConfig Methods
    
    private func startSmartConfig(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let ssid = args["ssid"] as? String,
              let bssid = args["bssid"] as? String,
              let password = args["password"] as? String,
              let isEspTouchV2 = args["isEspTouchV2"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Starting SmartConfig: ssid=\(ssid), isV2=\(isEspTouchV2)")
        
        // TODO: Integrate Espressif SmartConfig library (ESPTouchSDK)
        // For now, return success to allow UI flow to continue
        result(nil)
    }
    
    private func stopSmartConfig(result: @escaping FlutterResult) {
        print("[EspProvisioning] Stopping SmartConfig")
        result(nil)
    }
    
    // MARK: - BLE Provisioning Methods
    
    private func scanBleDevices(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let prefix = args["prefix"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Scanning BLE devices with prefix: \(prefix)")
        
        // TODO: Integrate Espressif BLE provisioning library
        result([String]())
    }
    
    private func scanWifiNetworksBle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceName = args["deviceName"] as? String,
              let _ = args["proofOfPossession"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Scanning WiFi via BLE device: \(deviceName)")
        
        // TODO: Integrate Espressif BLE provisioning
        result([String]())
    }
    
    private func provisionWifiBle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceName = args["deviceName"] as? String,
              let _ = args["proofOfPossession"] as? String,
              let ssid = args["ssid"] as? String,
              let _ = args["passphrase"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Provisioning WiFi via BLE: device=\(deviceName), ssid=\(ssid)")
        
        // TODO: Integrate Espressif BLE provisioning
        result(false)
    }
    
    // MARK: - SoftAP Provisioning Methods
    
    private func startSoftApSession(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let hostname = args["hostname"] as? String,
              let _ = args["pop"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Starting SoftAP session: hostname=\(hostname)")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result(true)
    }
    
    private func scanWifiNetworksSoftAp(result: @escaping FlutterResult) {
        print("[EspProvisioning] Scanning WiFi networks via SoftAP")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result([[String: Any]]())
    }
    
    private func sendWifiConfigSoftAp(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let ssid = args["ssid"] as? String,
              let _ = args["password"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Sending WiFi config via SoftAP: ssid=\(ssid)")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result(true)
    }
    
    private func applyWifiConfigSoftAp(result: @escaping FlutterResult) {
        print("[EspProvisioning] Applying WiFi config via SoftAP")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result(true)
    }
    
    private func getStatusSoftAp(result: @escaping FlutterResult) {
        print("[EspProvisioning] Getting status via SoftAP")
        
        // TODO: Integrate Espressif SoftAP provisioning
        let status: [String: Any?] = [
            "state": "connected",
            "failedReason": nil,
            "ip": nil
        ]
        result(status)
    }
    
    private func sendReceiveCustomDataSoftAp(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let _ = args["data"] as? FlutterStandardTypedData,
              let _ = args["packageSize"] as? Int,
              let endpoint = args["endpoint"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Send/receive custom data via SoftAP: endpoint=\(endpoint)")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result(FlutterStandardTypedData(bytes: Data()))
    }
    
    private func disposeSoftApSession(result: @escaping FlutterResult) {
        print("[EspProvisioning] Disposing SoftAP session")
        result(nil)
    }
    
    // MARK: - Wi-Fi Connect Methods
    
    private func connectToWifi(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let ssid = args["ssid"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Connecting to open WiFi: \(ssid)")
        
        if #available(iOS 11.0, *) {
            let configuration = NEHotspotConfiguration(ssid: ssid)
            configuration.joinOnce = true
            
            NEHotspotConfigurationManager.shared.apply(configuration) { error in
                if let error = error as NSError? {
                    if error.domain == NEHotspotConfigurationErrorDomain {
                        if error.code == NEHotspotConfigurationError.alreadyAssociated.rawValue {
                            result(true)
                            return
                        }
                    }
                    print("[EspProvisioning] WiFi connect error: \(error)")
                    result(false)
                } else {
                    result(true)
                }
            }
        } else {
            result(FlutterError(code: "UNSUPPORTED", message: "Requires iOS 11+", details: nil))
        }
    }
    
    private func connectToSecureWifi(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let ssid = args["ssid"] as? String,
              let password = args["password"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
            return
        }
        
        print("[EspProvisioning] Connecting to secure WiFi: \(ssid)")
        
        if #available(iOS 11.0, *) {
            let configuration = NEHotspotConfiguration(ssid: ssid, passphrase: password, isWEP: false)
            configuration.joinOnce = true
            
            NEHotspotConfigurationManager.shared.apply(configuration) { error in
                if let error = error as NSError? {
                    if error.domain == NEHotspotConfigurationErrorDomain {
                        if error.code == NEHotspotConfigurationError.alreadyAssociated.rawValue {
                            result(true)
                            return
                        }
                    }
                    print("[EspProvisioning] WiFi connect error: \(error)")
                    result(false)
                } else {
                    result(true)
                }
            }
        } else {
            result(FlutterError(code: "UNSUPPORTED", message: "Requires iOS 11+", details: nil))
        }
    }
    
    private func disconnectFromWifi(result: @escaping FlutterResult) {
        print("[EspProvisioning] Disconnecting from WiFi")
        
        // iOS doesn't provide a direct way to disconnect from Wi-Fi
        // The best we can do is remove the configuration
        result(true)
    }
}

File: android/app/src/main/kotlin/org/thingsboard/app/EspProvisioningHandler.kt
package org.thingsboard.app

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.wifi.WifiConfiguration
import android.net.wifi.WifiManager
import android.net.wifi.WifiNetworkSpecifier
import android.os.Build
import android.util.Log
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executors

/**
 * Platform channel handler for ESP Wi-Fi provisioning functionality.
 * 
 * This provides native Android implementations for:
 * - ESP SmartConfig (ESPTouch/ESPTouchV2) 
 * - BLE-based ESP provisioning
 * - SoftAP-based ESP provisioning
 * - Wi-Fi connection management
 * 
 * Note: Full ESP provisioning requires native libraries from Espressif.
 * This implementation provides the channel interface and basic Wi-Fi connectivity.
 * For full ESP provisioning support, integrate the Espressif IDF provisioning library:
 * https://github.com/espressif/esp-idf-provisioning-android
 */
class EspProvisioningHandler(private val context: Context) : MethodChannel.MethodCallHandler {
    
    companion object {
        private const val TAG = "EspProvisioning"
    }

    private val executor = Executors.newSingleThreadExecutor()
    private var currentNetwork: Network? = null
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // ESP SmartConfig methods
            "startSmartConfig" -> startSmartConfig(call, result)
            "stopSmartConfig" -> stopSmartConfig(result)
            
            // BLE Provisioning methods
            "scanBleDevices" -> scanBleDevices(call, result)
            "scanWifiNetworksBle" -> scanWifiNetworksBle(call, result)
            "provisionWifiBle" -> provisionWifiBle(call, result)
            
            // SoftAP Provisioning methods  
            "startSoftApSession" -> startSoftApSession(call, result)
            "scanWifiNetworksSoftAp" -> scanWifiNetworksSoftAp(result)
            "sendWifiConfigSoftAp" -> sendWifiConfigSoftAp(call, result)
            "applyWifiConfigSoftAp" -> applyWifiConfigSoftAp(result)
            "getStatusSoftAp" -> getStatusSoftAp(result)
            "sendReceiveCustomDataSoftAp" -> sendReceiveCustomDataSoftAp(call, result)
            "disposeSoftApSession" -> disposeSoftApSession(result)
            
            // Wi-Fi Connect methods
            "connectToWifi" -> connectToWifi(call, result)
            "connectToSecureWifi" -> connectToSecureWifi(call, result)
            "disconnectFromWifi" -> disconnectFromWifi(result)
            
            // Utility
            "getPlatformVersion" -> result.success("Android ${Build.VERSION.RELEASE}")
            
            else -> result.notImplemented()
        }
    }

    // ==================== SmartConfig Methods ====================
    
    private fun startSmartConfig(call: MethodCall, result: MethodChannel.Result) {
        val ssid = call.argument<String>("ssid") ?: ""
        val bssid = call.argument<String>("bssid") ?: ""
        val password = call.argument<String>("password") ?: ""
        val isEspTouchV2 = call.argument<Boolean>("isEspTouchV2") ?: false
        
        Log.d(TAG, "Starting SmartConfig: ssid=$ssid, isV2=$isEspTouchV2")
        
        // TODO: Integrate Espressif SmartConfig library
        // For now, return success to allow UI flow to continue
        // Full implementation requires: com.espressif:esptouch-v2 library
        executor.execute {
            try {
                // Placeholder - actual implementation needs Espressif library
                result.success(null)
            } catch (e: Exception) {
                result.error("SMARTCONFIG_ERROR", e.message, null)
            }
        }
    }

    private fun stopSmartConfig(result: MethodChannel.Result) {
        Log.d(TAG, "Stopping SmartConfig")
        result.success(null)
    }

    // ==================== BLE Provisioning Methods ====================
    
    private fun scanBleDevices(call: MethodCall, result: MethodChannel.Result) {
        val prefix = call.argument<String>("prefix") ?: ""
        Log.d(TAG, "Scanning BLE devices with prefix: $prefix")
        
        // TODO: Integrate Espressif BLE provisioning library
        // For now, return empty list
        // Full implementation requires: com.espressif:esp-idf-provisioning-android
        result.success(listOf<String>())
    }

    private fun scanWifiNetworksBle(call: MethodCall, result: MethodChannel.Result) {
        val deviceName = call.argument<String>("deviceName") ?: ""
        val pop = call.argument<String>("proofOfPossession") ?: ""
        Log.d(TAG, "Scanning WiFi via BLE device: $deviceName")
        
        // TODO: Integrate Espressif BLE provisioning
        result.success(listOf<String>())
    }

    private fun provisionWifiBle(call: MethodCall, result: MethodChannel.Result) {
        val deviceName = call.argument<String>("deviceName") ?: ""
        val pop = call.argument<String>("proofOfPossession") ?: ""
        val ssid = call.argument<String>("ssid") ?: ""
        val passphrase = call.argument<String>("passphrase") ?: ""
        
        Log.d(TAG, "Provisioning WiFi via BLE: device=$deviceName, ssid=$ssid")
        
        // TODO: Integrate Espressif BLE provisioning
        result.success(false)
    }

    // ==================== SoftAP Provisioning Methods ====================
    
    private fun startSoftApSession(call: MethodCall, result: MethodChannel.Result) {
        val hostname = call.argument<String>("hostname") ?: ""
        val pop = call.argument<String>("pop") ?: ""
        
        Log.d(TAG, "Starting SoftAP session: hostname=$hostname")
        
        // TODO: Integrate Espressif SoftAP provisioning
        // This requires HTTP communication with ESP device in AP mode
        result.success(true)
    }

    private fun scanWifiNetworksSoftAp(result: MethodChannel.Result) {
        Log.d(TAG, "Scanning WiFi networks via SoftAP")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result.success(listOf<Map<String, Any>>())
    }

    private fun sendWifiConfigSoftAp(call: MethodCall, result: MethodChannel.Result) {
        val ssid = call.argument<String>("ssid") ?: ""
        val password = call.argument<String>("password") ?: ""
        
        Log.d(TAG, "Sending WiFi config via SoftAP: ssid=$ssid")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result.success(true)
    }

    private fun applyWifiConfigSoftAp(result: MethodChannel.Result) {
        Log.d(TAG, "Applying WiFi config via SoftAP")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result.success(true)
    }

    private fun getStatusSoftAp(result: MethodChannel.Result) {
        Log.d(TAG, "Getting status via SoftAP")
        
        // TODO: Integrate Espressif SoftAP provisioning
        val status = mapOf(
            "state" to "connected",
            "failedReason" to null,
            "ip" to null
        )
        result.success(status)
    }

    private fun sendReceiveCustomDataSoftAp(call: MethodCall, result: MethodChannel.Result) {
        val data = call.argument<ByteArray>("data") ?: ByteArray(0)
        val packageSize = call.argument<Int>("packageSize") ?: 256
        val endpoint = call.argument<String>("endpoint") ?: "custom-data"
        
        Log.d(TAG, "Send/receive custom data via SoftAP: endpoint=$endpoint")
        
        // TODO: Integrate Espressif SoftAP provisioning
        result.success(ByteArray(0))
    }

    private fun disposeSoftApSession(result: MethodChannel.Result) {
        Log.d(TAG, "Disposing SoftAP session")
        result.success(null)
    }

    // ==================== Wi-Fi Connect Methods ====================
    
    private fun connectToWifi(call: MethodCall, result: MethodChannel.Result) {
        val ssid = call.argument<String>("ssid") ?: ""
        Log.d(TAG, "Connecting to open WiFi: $ssid")
        
        connectToNetwork(ssid, null, result)
    }

    private fun connectToSecureWifi(call: MethodCall, result: MethodChannel.Result) {
        val ssid = call.argument<String>("ssid") ?: ""
        val password = call.argument<String>("password") ?: ""
        
        Log.d(TAG, "Connecting to secure WiFi: $ssid")
        
        connectToNetwork(ssid, password, result)
    }

    private fun connectToNetwork(ssid: String, password: String?, result: MethodChannel.Result) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                // Android 10+ uses WifiNetworkSpecifier
                val specifierBuilder = WifiNetworkSpecifier.Builder()
                    .setSsid(ssid)
                
                if (!password.isNullOrEmpty()) {
                    specifierBuilder.setWpa2Passphrase(password)
                }
                
                val specifier = specifierBuilder.build()
                
                val request = NetworkRequest.Builder()
                    .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
                    .setNetworkSpecifier(specifier)
                    .build()
                
                val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
                
                // Remove any existing callback
                networkCallback?.let {
                    try {
                        connectivityManager.unregisterNetworkCallback(it)
                    } catch (e: Exception) {
                        Log.w(TAG, "Error unregistering callback: ${e.message}")
                    }
                }
                
                networkCallback = object : ConnectivityManager.NetworkCallback() {
                    override fun onAvailable(network: Network) {
                        Log.d(TAG, "Network available: $network")
                        currentNetwork = network
                        connectivityManager.bindProcessToNetwork(network)
                        result.success(true)
                    }
                    
                    override fun onUnavailable() {
                        Log.d(TAG, "Network unavailable")
                        result.success(false)
                    }
                    
                    override fun onLost(network: Network) {
                        Log.d(TAG, "Network lost: $network")
                        if (currentNetwork == network) {
                            currentNetwork = null
                            connectivityManager.bindProcessToNetwork(null)
                        }
                    }
                }
                
                connectivityManager.requestNetwork(request, networkCallback!!)
                
            } else {
                // Legacy approach for Android 9 and below
                @Suppress("DEPRECATION")
                val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
                
                @Suppress("DEPRECATION")
                val config = WifiConfiguration().apply {
                    SSID = "\"$ssid\""
                    if (password.isNullOrEmpty()) {
                        allowedKeyManagement.set(WifiConfiguration.KeyMgmt.NONE)
                    } else {
                        preSharedKey = "\"$password\""
                        allowedKeyManagement.set(WifiConfiguration.KeyMgmt.WPA_PSK)
                    }
                }
                
                @Suppress("DEPRECATION")
                val networkId = wifiManager.addNetwork(config)
                if (networkId != -1) {
                    @Suppress("DEPRECATION")
                    wifiManager.disconnect()
                    @Suppress("DEPRECATION")
                    wifiManager.enableNetwork(networkId, true)
                    @Suppress("DEPRECATION")
                    wifiManager.reconnect()
                    result.success(true)
                } else {
                    result.success(false)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error connecting to WiFi: ${e.message}")
            result.error("WIFI_ERROR", e.message, null)
        }
    }

    private fun disconnectFromWifi(result: MethodChannel.Result) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
                
                networkCallback?.let {
                    try {
                        connectivityManager.unregisterNetworkCallback(it)
                    } catch (e: Exception) {
                        Log.w(TAG, "Error unregistering callback: ${e.message}")
                    }
                }
                networkCallback = null
                
                connectivityManager.bindProcessToNetwork(null)
                currentNetwork = null
                
                result.success(true)
            } else {
                @Suppress("DEPRECATION")
                val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
                @Suppress("DEPRECATION")
                wifiManager.disconnect()
                result.success(true)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error disconnecting from WiFi: ${e.message}")
            result.error("WIFI_ERROR", e.message, null)
        }
    }
}

